src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -548,6 +548,10 @@ static void on_resolver_result_changed_locked(grpc_exec_ctx* exec_ctx,
       GRPC_RESOLVER_UNREF(exec_ctx, chand->resolver, "channel");
       chand->resolver = nullptr;
     }
+    if (chand->lb_policy != nullptr) {
```

Comment:
There are two cases in which we get to this point, as per the condition on line 539:
1. The resolver scheduled the closure with an error instead of returning results in `chand->resolver_result`.
2. The resolver scheduled the closure without an error and with results in `chand->resolver_result` but we were shut down (setting `chand->resolver` to null) by the time we got here.

In case 1, this change is unnecessary, because the condition on line 387 will be false, so we will never set `new_lb_policy`, and the code on lines 523-536 will ensure that we have already set `chand->lb_policy` to null.

In case 2, you're right that there is a potential problem here, but this fix is sub-optimal, because we will wind up creating a new LB policy which we then immediately unref.  Instead, I suggest changing line 387 to also check `&& chand->resolver != nullptr`, so that we avoid creating a new LB policy to begin with.  However, note that if `chand->resolver_result` is non-null, we still need to free it.

Reply:
> Since we are talking about avoiding deletion right after creation, should we actually check whether the client channel is shutting down at the beginning of `on_resolver_result_changed_locked()`? If the client channel is shutting down, we don't need to create the LB policy, or swap in the new values into chand. And the logic in `on_resolver_result_changed_locked()` would be easier to understand and deal with?

While it's true that we don't need to swap the new values into `chand` in this case, I think we do still need to unref the previous values.  This means that most of the code on lines 492-536 needs to be executed in both the shutdown and non-shutdown cases; the only distinction between them is that in the shutdown case, the new values we swap in are null.  It doesn't really make sense to duplicate that code; instead, the idea here is that in the shutdown case, the block on lines 387-484 is not executed, which leaves all of the new values initialized to null before we run the code to swap out the values.

(Also, resetting `chand->lb_policy` in particular is tricky, since we want to continue using the existing LB policy object if the LB policy name has not changed, and there are a couple of places in this function where we need to know whether or not we've created a new LB policy or are continuing to use the existing one.)

I can completely understand why you suggested this.  In fact, I've had this same thought myself on a few different occasions, but every time I look at doing this, it seems like it wouldn't actually help make the logic clearer.

> Another question is how we know the client channel is shutting down. The current code uses `error != GRPC_ERROR_NONE || chand->resolver == nullptr` to check and this appears at multiple locations.  I suspect we can check that at the beginning of `on_resolver_result_changed_locked()` and set the channel state into `SHUTDOWN` if needed, then use `grpc_connectivity_state_get()` afterwards. 

That condition only appears in two places.  The rule of three suggests that this isn't worth refactoring at this point.

Also, note that `grpc_connectivity_state_get()` does an atomic operation, so it's better for performance to use local variables here.

Reply:
I see. Change is added into the original PR #12829. Closing this.